### PR TITLE
Implement waiting for Connect My Computer node to join cluster

### PR DESCRIPTION
**Pull request by ravicious**
https://github.com/gravitational/teleport/blob/master/rfd/0133-connect-my-computer.md#determining-a-successful-launch

In short, after we start a Teleport agent in Connect, we want to wait until the node successfully joins the cluster so that we can change the UI state from "starting" to "running".

This is implemented by an RPC in tsh daemon which reads host UUID file from disk and then sets up a watcher to wait for `OpPut` of the node. Most of the work is done by `NodeJoinWait` in the `connectmycomputer` package. When the Electron app wants to wait until the node join, it calls this RPC.

Best reviewed commit-by-commit.

---

If you'd like to test it on a real cluster, then you need to checkout `ravicious/launch-js`, [open the config file](https://goteleport.com/docs/connect-your-client/teleport-connect/#configuration), set `feature.connectMyComputer` to `true`, restart the app and then click the laptop icon in the upper right of the view of all cluster resources.

The setup and subsequent launches should show a success state only once we detect that the node has joined the cluster.

### CODE REVIEW

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ravicious) [inline lib/teleterm/services/connectmycomputer/connectmycomputer.go:421]: Is there an easy way to parallelize both `GetNode` and `waitForOpPut` so that instead of waiting for a response from `GetNode` before processing events, we do both at the same time and return whatever is faster?

I wanted to use goroutines and a buffered channel to send the server over it, but I don't know how to handle errors when channels come into play. I'd like the semantics around errors to remain the same.

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ravicious) [inline lib/teleterm/services/connectmycomputer/connectmycomputer.go:391]: I couldn't use `utils.ReadHostUUID` because it returns `NotFound` both for when the file doesn't exist and when the host UUID in the file is empty.

I need to be able to distinguish between those two cases to enter the for loop again if the file was not created yet.

File: integration/teleterm_test.go
Comment (ibeckermayer) [inline integration/teleterm_test.go:719]: unchecked error

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ibeckermayer) [inline lib/teleterm/services/connectmycomputer/connectmycomputer.go:368]: Is there some way to create this path that ensures it stays consistent with whatever piece of code creates the file, which presumably needs to construct the same path?

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ibeckermayer) [reply lib/teleterm/services/connectmycomputer/connectmycomputer.go:391]: Worth an inline comment

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ibeckermayer) [inline lib/teleterm/services/connectmycomputer/connectmycomputer.go:375]: Should we have some sort of timeout here so we don't loop endlessly in the case that there's a problem creating the file?

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ibeckermayer) [inline lib/teleterm/services/connectmycomputer/connectmycomputer.go:421]: Wrong commit to comment, but I'm not seeing where we set a timeout for `waitForOpPut`.

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ravicious) [reply lib/teleterm/services/connectmycomputer/connectmycomputer.go:375]: To address both of your comments about timeouts, I didn't add any in the Go code because I expect the client to set for the whole request. The frontend app is going to abort this request after 20 seconds.

However, as you said, both operations essentially block forever if an error happens somewhere else. I suppose a timeout is warranted after all because of that, so I'm going to add a 1 minute timeout to `NodeJoinWait.Run`.

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ravicious) [reply lib/teleterm/services/connectmycomputer/connectmycomputer.go:435]: I won't be parallelizing this in the end. The cost of not doing this is very small. And from what I've seen on Slack, any attempt at making it so would make the code overall much more complex.

**Comment by gzdunek** [comment]
For some reason, after completing the setup, `WaitForConnectMyComputerNodeJoin` doesn't return the hostname value in labels:
```
[tshd] info: receive: /teleport.lib.teleterm.v1.TerminalService/WaitForConnectMyComputerNodeJoin -> ({"server":{"uri":"/clusters/mercury.cloud.gravitational.io/servers/fd4c100e-47d4-489d-a0f0-edfcc9eaf086","tunnel":true,"name":"fd4c100e-47d4-489d-a0f0-edfcc9eaf086","hostname":"mbp.home","addr":"","labelsList":{"0":{"name":"hostname","value":""},"1":{"name":"teleport.dev/connect-my-computer/owner","value":"grzegorz.zdunek@goteleport.com"}}}})
```

However, on subsequent calls it works. 

Does it happen to you too?

**Comment by ravicious** [comment]
@gzdunek That's probably because the hostname label is dynamic and it's not yet set when the first `OpPut` event comes in. On subsequent calls, the RPC returns the node through `GetNode` rather than from the event.

I suspected that returning the resource from the event might bite me, but not in this way.

We could certainly attempt to refetch the node after receiving the event rather than returning it straight from the event. What do you think?

**Comment by gzdunek** [comment]
>We could certainly attempt to refetch the node after receiving the event rather than returning it straight from the event. What do you think?

Yeah, let's do this.

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ibeckermayer) [reply lib/teleterm/services/connectmycomputer/connectmycomputer.go]: I figured that the file was being created somewhere with Go, and so was wondering if there might be some way to marry the code, however I see that that's not the case. Your refactor seems fine nevertheless.

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ibeckermayer) [inline lib/teleterm/services/connectmycomputer/connectmycomputer.go:356]:
```suggestion
	// The Electron app aborts this request after a timeout, but we set a timeout here as well for
	// to ensure that the request can't hang forever if there's a bug in the request abort logic.
```

Two things to consider:

1. Perhaps move this `timeoutCtx` up into [the endpoint itself](https://github.com/gravitational/teleport/blob/e384604811d9c2f73231efd0a16133d5dbcb993d/lib/teleterm/apiserver/handler/handler_connectmycomputer.go#L59-L73). That way it's clear that the whole endpoint just times out after a minute.
2. I'm not sure how big of a change this would be, but since we're adding timeout logic here anyways, there may be a case for just removing it from the frontend code and using this logic exclusively instead, or having the frontend pass the timeout as a request parameter.

Comment (ravicious) [reply lib/teleterm/services/connectmycomputer/connectmycomputer.go:356]: > 1. Perhaps move this `timeoutCtx` up into [the endpoint itself](https://github.com/gravitational/teleport/blob/e384604811d9c2f73231efd0a16133d5dbcb993d/lib/teleterm/apiserver/handler/handler_connectmycomputer.go#L59-L73). That way it's clear that the whole endpoint just times out after a minute.

Yeah, I wasn't sure if it'd be best to add a timeout tightly around the operation that actually blocks vs for the whole endpoint.

Since those individual services technically could serve as building blocks for other requests, I guess it's best to add the timeout on the endpoint level.

> 2\. I'm not sure how big of a change this would be, but since we're adding timeout logic here anyways, there may be a case for just removing it from the frontend code and using this logic exclusively instead, or having the frontend pass the timeout as a request parameter.

Now that I think about it, under normal circumstances this would be the way to go. However, the UI needs to be able to distinguish between a timeout and other errors. The JS gRPC client encodes that information as custom properties on the error object. gRPC errors in our app pass through [the Electron context bridge](https://www.electronjs.org/docs/latest/api/context-bridge) which strips all custom properties from Error objects, meaning we'd only have the error message to distinguish between a timeout or not.

So until we revamp how those errors are passed, I think I'm going to keep current implementation as is with regards to the timeout.

I'm still not sure if an endpoint should trouble itself with the deadline, perhaps it should rather return `BadParameter` if no deadline was set on the gRPC request? I'd rather use built-in mechanisms to handle stuff like this vs building my own on top of gRPC.

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ravicious) [inline lib/teleterm/services/connectmycomputer/connectmycomputer.go:426]: @gzdunek Regarding the problem with a blank hostname, it's not going to be as simple as refetching the node after `OpPut` comes in.

I just implemented this and even when connecting to a cluster in our cloud, the label update doesn't happen fast enough. We could wait for the hostname label to get filled out, but this would complicate the implementation and I don't think it's entirely necessary.

What if we used some smoke and mirrors here? If there's a hostname label (and there will be one by default) and it's empty, we fill it out with what `os.Hostname` reports.

Comment (gzdunek) [reply lib/teleterm/services/connectmycomputer/connectmycomputer.go:426]: >We could wait for the hostname label to get filled out, but this would complicate the implementation and I don't think it's entirely necessary.

Yeah, it is not worth it.

>If there's a hostname label (and there will be one by default) and it's empty, we fill it out with what os.Hostname reports.

Works for me!

**Review by gzdunek — APPROVED** [review]

**Comment by ravicious** [comment]
Ping @ibeckermayer, we have two other PRs stacked on this one so I'd like to merge it if possible. ;)

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ibeckermayer) [inline lib/teleterm/services/connectmycomputer/connectmycomputer.go:365]: What this comment is pointing out isn't quite clear to me (granted it could just be a function of the fact that I'm unfamiliar with this part of the codebase).

**Review by ibeckermayer — APPROVED** [review]

File: lib/teleterm/services/connectmycomputer/connectmycomputer.go
Comment (ravicious) [reply lib/teleterm/services/connectmycomputer/connectmycomputer.go:365]: Oh, I think the comment doesn't make it clear that the config comes from `teleport node config` that we call during the setup of Connect My Computer.

I'll update the comment to my best ability and then merge this PR, but please let me know if the updated comment is any better.

**Comment by ravicious** [comment]
Merged #30910 into this PR with fast forward.

**Review by avatus — APPROVED** [review]

**Pull request merged**
